xds/internal/client/dump_test.go

```diff
@@ -0,0 +1,443 @@
+				Err:     nackErr,
+			},
+		},
+		Raw: clusterRaws[cdsTargets[0]],
+	}
+	wantDump[cdsTargets[1]] = UpdateWithMD{
+		MD:  UpdateMetadata{Version: testVersion},
+		Raw: clusterRaws[cdsTargets[1]],
+	}
+	compareDump(t, client.DumpCDS, nackVersion, wantDump)
+}
+
+func (s) TestEDSConfigDump(t *testing.T) {
+	const testVersion = "test-version-cds"
+	var (
+		edsTargets    = []string{"cluster-0", "cluster-1"}
+		localityNames = []string{"locality-0", "locality-1"}
+		addrs         = []string{"addr0:123", "addr1:456"}
+		endpointRaws  = make(map[string]*anypb.Any, len(edsTargets))
+	)
+
+	for i := range edsTargets {
+		clab0 := newClaBuilder(edsTargets[i], nil)
+		clab0.addLocality(localityNames[i], 1, 1, []string{addrs[i]}, nil)
+		claT := clab0.Build()
+
+		anyT, err := ptypes.MarshalAny(claT)
+		if err != nil {
+			t.Fatalf("failed to marshal proto to any: %v", err)
+		}
+		endpointRaws[edsTargets[i]] = anyT
+	}
+
+	client, err := newWithConfig(clientOpts(testXDSServer, false))
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+	defer client.Close()
+
+	// Expected unknown.
+	compareDump(t, client.DumpEDS, "", map[string]UpdateWithMD{})
+
+	wantRequested := make(map[string]UpdateWithMD)
+	for _, n := range edsTargets {
+		cancel := client.WatchEndpoints(n, func(update EndpointsUpdate, err error) {})
+		defer cancel()
+		wantRequested[n] = UpdateWithMD{MD: UpdateMetadata{Status: ServiceStatusRequested}}
+	}
+	// Expected requested.
+	compareDump(t, client.DumpEDS, "", wantRequested)
+
+	update0 := make(map[string]EndpointsUpdate)
+	want0 := make(map[string]UpdateWithMD)
+	for n, r := range endpointRaws {
+		update0[n] = EndpointsUpdate{Raw: r}
+		want0[n] = UpdateWithMD{
+			MD:  UpdateMetadata{Version: testVersion},
+			Raw: r,
+		}
+	}
+	client.NewEndpoints(update0, UpdateMetadata{Version: testVersion})
+
+	// Expect ACK.
+	compareDump(t, client.DumpEDS, testVersion, want0)
+
+	const nackVersion = "eds-version-nack"
+	var nackErr = fmt.Errorf("eds nack error")
+	client.NewEndpoints(
+		map[string]EndpointsUpdate{
+			edsTargets[0]: {},
+		},
+		UpdateMetadata{
+			ErrState: &UpdateErrorMetadata{
+				Version: nackVersion,
+				Err:     nackErr,
+			},
+		},
+	)
+
+	// Expect NACK for [0], but old ACK for [1].
+	wantDump := make(map[string]UpdateWithMD)
+	// Though resource 0 was NACKed, the dump should show the previous ACKed raw
+	// message, as well as the NACK error.
+	wantDump[edsTargets[0]] = UpdateWithMD{
+		MD: UpdateMetadata{
+			Version: testVersion,
+			ErrState: &UpdateErrorMetadata{
+				Version: nackVersion,
+				Err:     nackErr,
+			},
+		},
+		Raw: endpointRaws[edsTargets[0]],
+	}
+	wantDump[edsTargets[1]] = UpdateWithMD{
+		MD:  UpdateMetadata{Version: testVersion},
+		Raw: endpointRaws[edsTargets[1]],
+	}
+	compareDump(t, client.DumpEDS, nackVersion, wantDump)
+}
+
+func compareDump(t *testing.T, dumpFunc func() (string, map[string]UpdateWithMD), wantVersion string, wantDump interface{}) {
```

Comment:
Test helper is a function that performs a setup or teardown task, such as constructing an input message, that does not depend on the code under test. See: go/go-style/decisions#mark-test-helpers

This function should ideally be changed to return an error and let the test handle it.

Reply:
Done